<!DOCTYPE HTML>
<html lang="en" class="light sidebar-visible" dir="ltr">
    <head>
        <!-- Book generated using mdBook -->
        <meta charset="UTF-8">
        <title>Extended subcommand help - Modkit</title>


        <!-- Custom HTML head -->

        <meta name="description" content="">
        <meta name="viewport" content="width=device-width, initial-scale=1">
        <meta name="theme-color" content="#ffffff">

        <link rel="icon" href="favicon-de23e50b.svg">
        <link rel="shortcut icon" href="favicon-8114d1fc.png">
        <link rel="stylesheet" href="css/variables-8adf115d.css">
        <link rel="stylesheet" href="css/general-2459343d.css">
        <link rel="stylesheet" href="css/chrome-ae938929.css">
        <link rel="stylesheet" href="css/print-9e4910d8.css" media="print">

        <!-- Fonts -->
        <link rel="stylesheet" href="fonts/fonts-9644e21d.css">

        <!-- Highlight.js Stylesheets -->
        <link rel="stylesheet" id="mdbook-highlight-css" href="highlight-493f70e1.css">
        <link rel="stylesheet" id="mdbook-tomorrow-night-css" href="tomorrow-night-4c0ae647.css">
        <link rel="stylesheet" id="mdbook-ayu-highlight-css" href="ayu-highlight-3fdfc3ac.css">

        <!-- Custom theme stylesheets -->
        <link rel="stylesheet" href="custom-03c054ce.css">

        <!-- MathJax -->
        <script async src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML"></script>

        <!-- Provide site root and default themes to javascript -->
        <script>
            const path_to_root = "";
            const default_light_theme = "light";
            const default_dark_theme = "navy";
            window.path_to_searchindex_js = "searchindex-837351bc.js";
        </script>
        <!-- Start loading toc.js asap -->
        <script src="toc-2cf11e43.js"></script>
    </head>
    <body>
    <div id="mdbook-help-container">
        <div id="mdbook-help-popup">
            <h2 class="mdbook-help-title">Keyboard shortcuts</h2>
            <div>
                <p>Press <kbd>←</kbd> or <kbd>→</kbd> to navigate between chapters</p>
                <p>Press <kbd>S</kbd> or <kbd>/</kbd> to search in the book</p>
                <p>Press <kbd>?</kbd> to show this help</p>
                <p>Press <kbd>Esc</kbd> to hide this help</p>
            </div>
        </div>
    </div>
    <div id="mdbook-body-container">
        <!-- Work around some values being stored in localStorage wrapped in quotes -->
        <script>
            try {
                let theme = localStorage.getItem('mdbook-theme');
                let sidebar = localStorage.getItem('mdbook-sidebar');

                if (theme.startsWith('"') && theme.endsWith('"')) {
                    localStorage.setItem('mdbook-theme', theme.slice(1, theme.length - 1));
                }

                if (sidebar.startsWith('"') && sidebar.endsWith('"')) {
                    localStorage.setItem('mdbook-sidebar', sidebar.slice(1, sidebar.length - 1));
                }
            } catch (e) { }
        </script>

        <!-- Set the theme before any content is loaded, prevents flash -->
        <script>
            const default_theme = window.matchMedia("(prefers-color-scheme: dark)").matches ? default_dark_theme : default_light_theme;
            let theme;
            try { theme = localStorage.getItem('mdbook-theme'); } catch(e) { }
            if (theme === null || theme === undefined) { theme = default_theme; }
            const html = document.documentElement;
            html.classList.remove('light')
            html.classList.add(theme);
            html.classList.add("js");
        </script>

        <input type="checkbox" id="mdbook-sidebar-toggle-anchor" class="hidden">

        <!-- Hide / unhide sidebar before it is displayed -->
        <script>
            let sidebar = null;
            const sidebar_toggle = document.getElementById("mdbook-sidebar-toggle-anchor");
            if (document.body.clientWidth >= 1080) {
                try { sidebar = localStorage.getItem('mdbook-sidebar'); } catch(e) { }
                sidebar = sidebar || 'visible';
            } else {
                sidebar = 'hidden';
                sidebar_toggle.checked = false;
            }
            if (sidebar === 'visible') {
                sidebar_toggle.checked = true;
            } else {
                html.classList.remove('sidebar-visible');
            }
        </script>

        <nav id="mdbook-sidebar" class="sidebar" aria-label="Table of contents">
            <!-- populated by js -->
            <mdbook-sidebar-scrollbox class="sidebar-scrollbox"></mdbook-sidebar-scrollbox>
            <noscript>
                <iframe class="sidebar-iframe-outer" src="toc.html"></iframe>
            </noscript>
            <div id="mdbook-sidebar-resize-handle" class="sidebar-resize-handle">
                <div class="sidebar-resize-indicator"></div>
            </div>
        </nav>

        <div id="mdbook-page-wrapper" class="page-wrapper">

            <div class="page">
                <div id="mdbook-menu-bar-hover-placeholder"></div>
                <div id="mdbook-menu-bar" class="menu-bar sticky">
                    <div class="left-buttons">
                        <label id="mdbook-sidebar-toggle" class="icon-button" for="mdbook-sidebar-toggle-anchor" title="Toggle Table of Contents" aria-label="Toggle Table of Contents" aria-controls="mdbook-sidebar">
                            <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 448 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M0 96C0 78.3 14.3 64 32 64H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32C14.3 128 0 113.7 0 96zM0 256c0-17.7 14.3-32 32-32H416c17.7 0 32 14.3 32 32s-14.3 32-32 32H32c-17.7 0-32-14.3-32-32zM448 416c0 17.7-14.3 32-32 32H32c-17.7 0-32-14.3-32-32s14.3-32 32-32H416c17.7 0 32 14.3 32 32z"/></svg></span>
                        </label>
                        <button id="mdbook-theme-toggle" class="icon-button" type="button" title="Change theme" aria-label="Change theme" aria-haspopup="true" aria-expanded="false" aria-controls="mdbook-theme-list">
                            <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M371.3 367.1c27.3-3.9 51.9-19.4 67.2-42.9L600.2 74.1c12.6-19.5 9.4-45.3-7.6-61.2S549.7-4.4 531.1 9.6L294.4 187.2c-24 18-38.2 46.1-38.4 76.1L371.3 367.1zm-19.6 25.4l-116-104.4C175.9 290.3 128 339.6 128 400c0 3.9 .2 7.8 .6 11.6c1.8 17.5-10.2 36.4-27.8 36.4H96c-17.7 0-32 14.3-32 32s14.3 32 32 32H240c61.9 0 112-50.1 112-112c0-2.5-.1-5-.2-7.5z"/></svg></span>
                        </button>
                        <ul id="mdbook-theme-list" class="theme-popup" aria-label="Themes" role="menu">
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-default_theme">Auto</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-light">Light</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-rust">Rust</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-coal">Coal</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-navy">Navy</button></li>
                            <li role="none"><button role="menuitem" class="theme" id="mdbook-theme-ayu">Ayu</button></li>
                        </ul>
                        <button id="mdbook-search-toggle" class="icon-button" type="button" title="Search (`/`)" aria-label="Toggle Searchbar" aria-expanded="false" aria-keyshortcuts="/ s" aria-controls="mdbook-searchbar">
                            <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M416 208c0 45.9-14.9 88.3-40 122.7L502.6 457.4c12.5 12.5 12.5 32.8 0 45.3s-32.8 12.5-45.3 0L330.7 376c-34.4 25.2-76.8 40-122.7 40C93.1 416 0 322.9 0 208S93.1 0 208 0S416 93.1 416 208zM208 352c79.5 0 144-64.5 144-144s-64.5-144-144-144S64 128.5 64 208s64.5 144 144 144z"/></svg></span>
                        </button>
                    </div>

                    <h1 class="menu-title">Modkit</h1>

                    <div class="right-buttons">
                        <a href="print.html" title="Print this book" aria-label="Print this book">
                            <span class=fa-svg id="print-button"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M128 0C92.7 0 64 28.7 64 64v96h64V64H354.7L384 93.3V160h64V93.3c0-17-6.7-33.3-18.7-45.3L400 18.7C388 6.7 371.7 0 354.7 0H128zM384 352v32 64H128V384 368 352H384zm64 32h32c17.7 0 32-14.3 32-32V256c0-35.3-28.7-64-64-64H64c-35.3 0-64 28.7-64 64v96c0 17.7 14.3 32 32 32H64v64c0 35.3 28.7 64 64 64H384c35.3 0 64-28.7 64-64V384zm-16-88c-13.3 0-24-10.7-24-24s10.7-24 24-24s24 10.7 24 24s-10.7 24-24 24z"/></svg></span>
                        </a>

                    </div>
                </div>

                <div id="mdbook-search-wrapper" class="hidden">
                    <form id="mdbook-searchbar-outer" class="searchbar-outer">
                        <div class="search-wrapper">
                            <input type="search" id="mdbook-searchbar" name="searchbar" placeholder="Search this book ..." aria-controls="mdbook-searchresults-outer" aria-describedby="searchresults-header">
                            <div class="spinner-wrapper">
                                <span class=fa-svg id="fa-spin"><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M304 48c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zm0 416c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zM48 304c26.5 0 48-21.5 48-48s-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48zm464-48c0-26.5-21.5-48-48-48s-48 21.5-48 48s21.5 48 48 48s48-21.5 48-48zM142.9 437c18.7-18.7 18.7-49.1 0-67.9s-49.1-18.7-67.9 0s-18.7 49.1 0 67.9s49.1 18.7 67.9 0zm0-294.2c18.7-18.7 18.7-49.1 0-67.9S93.7 56.2 75 75s-18.7 49.1 0 67.9s49.1 18.7 67.9 0zM369.1 437c18.7 18.7 49.1 18.7 67.9 0s18.7-49.1 0-67.9s-49.1-18.7-67.9 0s-18.7 49.1 0 67.9z"/></svg></span>
                            </div>
                        </div>
                    </form>
                    <div id="mdbook-searchresults-outer" class="searchresults-outer hidden">
                        <div id="mdbook-searchresults-header" class="searchresults-header"></div>
                        <ul id="mdbook-searchresults">
                        </ul>
                    </div>
                </div>

                <!-- Apply ARIA attributes after the sidebar and the sidebar toggle button are added to the DOM -->
                <script>
                    document.getElementById('mdbook-sidebar-toggle').setAttribute('aria-expanded', sidebar === 'visible');
                    document.getElementById('mdbook-sidebar').setAttribute('aria-hidden', sidebar !== 'visible');
                    Array.from(document.querySelectorAll('#mdbook-sidebar a')).forEach(function(link) {
                        link.setAttribute('tabIndex', sidebar === 'visible' ? 0 : -1);
                    });
                </script>

                <div id="mdbook-content" class="content">
                    <main>
                        <h1 id="modkit-subcommand-documentation"><a class="header" href="#modkit-subcommand-documentation">modkit, subcommand documentation</a></h1>
<p>The goal of <code>modkit</code> is to enable best-practices manipulation of BAM files containing
modified base information (modBAMs). The various sub-commands and tools available in
<code>modkit</code> are described below.  This information can be obtained by invoking the long help
(<code>--help</code>) for each command.</p>
<blockquote>
<p>Advanced usage information.</p>
</blockquote>
<pre><code class="language-text">Modkit is a bioinformatics tool for working with modified bases from Oxford
Nanopore

Usage: modkit &lt;COMMAND&gt;

Commands:
  pileup          Tabulates base modification calls across genomic positions.
                  This command produces a bedMethyl formatted file. Schema and
                  description of fields can be found in the README
  extract         Extract read-level base modification information from a modBAM
                  into a tab-separated values table
  dmr             Perform DMR test on a set of regions. Output a BED file of
                  regions with the score column indicating the magnitude of the
                  difference. Find the schema and description of fields can in
                  the README as well as a description of the model and method.
                  See subcommand help for additional details
  pileup-hemi     Tabulates double-stranded base modification patters (such as
                  hemi-methylation) across genomic motif positions. This command
                  produces a bedMethyl file, the schema can be found in the
                  online documentation
  validate        Validate results from a set of mod-BAM files and associated
                  BED files containing the ground truth modified base status at
                  reference positions
  motif           Various commands to search for, evaluate, or further regine
                  sequence motifs enriched for base modification. Also can
                  generate BED files of motif locations
  entropy         Use a mod-BAM to calculate methylation entropy over genomic
                  windows
  localize        Investigate patterns of base modifications, by aggregating
                  pileup counts "localized" around genomic features of interest
  stats           Calculate base modification levels over regions
  bedmethyl       Utilities to work with bedMethyl files
  modbam          Utilities to work with modBAM files
  open-chromatin  Identify regions of open chromatin based on exogenous 6mA
                  signal
  help            Print this message or the help of the given subcommand(s)

Options:
  -h, --help     Print help
  -V, --version  Print version
</code></pre>
<h2 id="pileup"><a class="header" href="#pileup">pileup</a></h2>
<pre><code class="language-text">Tabulates base modification calls across genomic positions. This command
produces a bedMethyl formatted file. Schema and description of fields can be
found in the README

Usage: modkit pileup [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BED&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available

  &lt;OUT_BED&gt;
          Output file to write results into. Specify "-" or "stdout" to direct
          output to stdout

Options:
      --duplex
          Specify that modBAM contains duplex modification calls and these
          should be used

      --modified-bases &lt;MODIFIED_BASES&gt;...
          

  -h, --help
          Print help (see a summary with '-h')

Output Options:
      --bgzf
          Output bgzf-compressed bedmethyl files suitable for Tabix-indexing

      --mixed-delim
          Output bedMethyl where the delimiter of columns past column 10 are
          space-delimited instead of tab-delimited. This option can be useful
          for some browsers and parsers that don't expect the extra columns of
          the bedMethyl format

      --header
          Output a header with the bedMethyl

      --prefix &lt;PREFIX&gt;
          Prefix to prepend on phased output file names. Without this option the
          files will be &lt;mod_code&gt;_&lt;strand&gt;.bedmethyl

      --phased
          Produce separate bedMethyl tables for haplotype-labeled modBAM
          records. Currently only supports diploid genomes. Produces
          hp1.bedmethyl, hp2.bedmethyl, and combined.bedmethyl files (optionally
          compressed). hp1.bedmethyl and hp2.bedmethyl contain counts for
          records with HP=1 and HP=2 tags, respectively. combined.bedmethyl
          contains counts for all modBAM records

Compute Options:
      --bgzf-threads &lt;BGZF_THREADS&gt;
          Specify the number of threads to dedicate to BGZF compression
          
          [default: 4]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently
          
          [default: 4]

  -s, --sampling-threads &lt;SAMPLING_THREADS&gt;
          Use this many threads when performing threshold estimation, setting
          this to a higher value can speed up execution

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead
          
          [default: 1000000]

      --queue-size &lt;QUEUE_SIZE&gt;
          Size of queue for writing records, default will be the number of
          threads

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended. (alias: log)

      --suppress-progress
          Hide the progress bar

Selection Options:
      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup.
          Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas
          are allowed

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to
          positions overlapping intervals in the file. (alias:
          include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads
          will be sampled evenly across aligned genome. If a region is
          specified, either with the --region option or the --sample-region
          option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand
          reads is sufficient to estimate the model output distribution and
          determine the filtering threshold
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold.
          In practice, 10-100 thousand reads is sufficient to estimate the model
          output distribution and determine the filtering threshold. See
          filtering.md for details on filtering

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is
          non-deterministic

Filtering Options:
      --no-filtering
          Do not perform any filtering, include all mod base calls in output.
          See filtering.md for details on filtering

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter
          threshold can be specified with by a decimal number (e.g. 0.75).
          Per-base thresholds can be specified by colon-separated values, for
          example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when
          estimating the threshold probability, can be larger than the pileup
          processing interval
          
          [default: 1000000]

Modified Base Options:
      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output pileup counts for only sequence motifs provided. The first
          argument should be the sequence motif and the second argument is the
          0-based offset to the base to pileup base modification counts for. For
          example: --motif CGCG 0 indicates to pileup counts for the first C on
          the top strand and the last C (complement to G) on the bottom strand.
          The --cpg argument is short hand for --motif CG 0.
          
          This argument can be passed multiple times. When more than one motif
          is used, the resulting output BED file will indicate the motif in the
          "name" field as &lt;mod_code&gt;,&lt;motif&gt;,&lt;offset&gt;. For example, given
          `--motif CGCG 2 --motif CG 0` there will be output lines with name
          fields such as "m,CG,0" and "m,CGCG,2". To use this option with
          `--combine-strands`, all motifs must be reverse-complement palindromic
          or an error will be raised.

      --cpg
          Only output counts at CpG motifs

  -r, --reference &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format. Required for motif (e.g. CpG)
          filtering, requires FAI fasta index to be pre-generated. (alias:
          'ref')

      --preload-references
          Preload the reference sequences, useful when working with many, short
          reference sequences such as a transcriptome

  -k, --mask
          Respect soft masking in the reference FASTA

      --combine-mods
          Combine base modification calls, all counts of modified bases are
          summed together. See collapse.md for details

      --combine-strands
          When performing motif analysis (such as CpG), sum the counts from the
          positive and negative strands into the counts for the positive strand
          position
</code></pre>
<h2 id="adjust-mods"><a class="header" href="#adjust-mods">adjust-mods</a></h2>
<pre><code class="language-text">Performs various operations on BAM files containing base modification
information, such as converting base modification codes and ignoring
modification calls. Produces a BAM output file

Usage: modkit adjust-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM file, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

  &lt;OUT_BAM&gt;
          File path to new BAM file to be created. Can be a path to a file or
          one of `-` or `stdin` to specify a stream from standard output

Options:
  -f, --ff
          Fast fail, stop processing at the first invalid sequence record.
          Default behavior is to continue and report failed/skipped records at
          the end

  -h, --help
          Print help (see a summary with '-h')

Output Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output debug logs to file at this path

      --output-sam
          Output SAM format instead of BAM

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Modified base code to ignore/remove, see
          https://samtools.github.io/hts-specs/SAMtags.pdf for details on the
          modified base codes

      --convert &lt;CONVERT&gt; &lt;CONVERT&gt;
          Convert one mod-tag to another, summing the probabilities together if
          the retained mod tag is already present

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Filter out any base modification call that isn't part of a basecall
          sequence motif. This argument can be passed multiple times. Format is
          &lt;motif_sequence&gt; &lt;offset&gt;. For example the argument to match CpG
          dinucleotides is `--motif CG 0`, or to match CG[5mC]G the argument
          would be `--motif CGCG 2`. Single bases can be used as motifs to keep
          only base modification calls for a specific primary base, for example
          `--motif C 0`

      --cpg
          Shorthand for --motif CG 0

      --discard-motifs
          Discard base modification calls that match the provided motifs
          (instead of keeping them)

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --filter-probs
          Filter out the lowest confidence base modification probabilities

      --mapped-only
          Only use base modification probabilities from bases that are aligned
          when estimating the filter threshold (i.e. ignore soft-clipped, and
          inserted bases)

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering
          threshold. If alignments are present reads will be sampled evenly
          across aligned genome. If a region is specified, either with the
          --region option or the --sample-region option, then reads will be
          sampled evenly across the region given. This option is useful for
          large BAM files. In practice, 10-50 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold
          
          [default: 10042]

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the
          threshold probability, can be larger than the pileup processing
          interval
          
          [default: 1000000]

Filtering Options:
  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global
          filter threshold can be specified with by a decimal number (e.g.
          0.75). Per-base thresholds can be specified by colon-separated values,
          for example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

Logging Options:
      --suppress-progress
          Hide the progress bar
</code></pre>
<h2 id="update-tags"><a class="header" href="#update-tags">update-tags</a></h2>
<pre><code class="language-text">Renames Mm/Ml to tags to MM/ML. Also allows changing the mode flag from silent
'.' to explicitly '?' or '.'

Usage: modkit update-tags [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;   BAM to update modified base tags in. Can be a path to a file or one
             of `-` or `stdin` to specify a stream from standard input
  &lt;OUT_BAM&gt;  File to new BAM file to be created or one of `-` or `stdin` to
             specify a stream from standard output

Options:
  -m, --mode &lt;MODE&gt;        Mode, change mode to this value, options {'explicit',
                           'implicit'}. See spec at:
                           https://samtools.github.io/hts-specs/SAMtags.pdf.
                           'explicit' ('?') means residues without modification
                           probabilities will not be assumed canonical or
                           modified. 'implicit' means residues without explicit
                           modification probabilities are assumed to be
                           canonical [possible values: explicit, implicit]
      --no-implicit-probs  Don't add implicit canonical calls. This flag is
                           important when converting from one of the implicit
                           modes ( `.` or `""`) to explicit mode (`?`). By
                           passing this flag, the bases without associated base
                           modification probabilities will not be assumed to be
                           canonical. No base modification probability will be
                           written for these bases, meaning there is no
                           information. The mode will automatically be set to
                           the explicit mode `?`
  -h, --help               Print help

Compute Options:
  -t, --threads &lt;THREADS&gt;  Number of threads to use [default: 4]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;  Output debug logs to file at this path

Output Options:
      --output-sam  Output SAM format instead of BAM
</code></pre>
<h2 id="sample-probs"><a class="header" href="#sample-probs">sample-probs</a></h2>
<pre><code class="language-text">Calculate an estimate of the base modification probability distribution

Usage: modkit sample-probs [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM with modified base tags. If a index is found reads will be
          sampled evenly across the length of the reference sequence. Can be a
          path to a file or one of `-` or `stdin` to specify a stream from
          standard input

Options:
  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          Only used when sampling probs from an indexed bam
          
          [default: 1000000]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --suppress-progress
          Hide the progress bar

Output Options:
  -p, --percentiles &lt;PERCENTILES&gt;
          Percentiles to calculate, a space separated list of floats
          
          [default: 0.1,0.5,0.9]

  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to deposit result tables into. Required for model
          probability histogram output

      --prefix &lt;PREFIX&gt;
          Label to prefix output files with

      --force
          Overwrite results if present

      --hist
          Output histogram of base modification prediction probabilities

      --dna-color &lt;PRIMARY_BASE_COLORS&gt; &lt;PRIMARY_BASE_COLORS&gt;
          Set colors of primary bases in histogram, should be RGB format, e.g.
          "#0000FF" is defailt for canonical cytosine

      --mod-color &lt;MOD_BASE_COLORS&gt; &lt;MOD_BASE_COLORS&gt;
          Set colors of modified bases in histogram, should be RGB format, e.g.
          "#FF00FF" is default for 5hmC

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting
          probabilities. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or
          &lt;chrom_name&gt;

      --include-bed &lt;INCLUDE_BED&gt;
          Only sample base modification probabilities that are aligned to the
          positions in this BED file. (alias: include-positions)

      --mapped-only
          Only use base modification probabilities that are aligned (i.e. ignore
          soft-clipped, and inserted bases)

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended
          when using a large BAM without an index. If an indexed BAM is
          provided, the reads will be sampled evenly over the length of the
          aligned reference. If a region is passed with the --region option,
          they will be sampled over the genomic region. Actual number of reads
          used may deviate slightly from this number
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of
          reads to sample, for example 0.1 will sample 1/10th of the reads

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds

  -s, --seed &lt;SEED&gt;
          Random seed for deterministic running, the default is
          non-deterministic, only used when no BAM index is provided
</code></pre>
<h2 id="summary"><a class="header" href="#summary">summary</a></h2>
<pre><code class="language-text">Summarize the mod tags present in a BAM and get basic statistics. The default
output is a totals table (designated by '#' lines) and a modification calls
table. Descriptions of the columns can be found in the README

Usage: modkit summary [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBam, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

Options:
  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          When using regions, interval chunk size in base pairs to process
          concurrently. Smaller interval chunk sizes will use less memory but
          incur more overhead
          
          [default: 1000000]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --suppress-progress
          Hide the progress bar

Output Options:
      --tsv
          Output summary as a tab-separated variables stdout instead of a table

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended
          when using a large BAM without an index. If an indexed BAM is
          provided, the reads will be sampled evenly over the length of the
          aligned reference. If a region is passed with the --region option,
          they will be sampled over the genomic region. Actual number of reads
          used may deviate slightly from this number
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of
          reads to sample when estimating the filter threshold. For example 0.1
          will sample 1/10th of the reads

      --no-sampling
          No sampling, use all the reads to calculate the filter thresholds and
          generating the summary

  -s, --seed &lt;SEED&gt;
          Sets a random seed for deterministic running (when using
          --sample-frac), the default is non-deterministic, only used when no
          BAM index is provided

Filtering Options:
      --no-filtering
          Do not perform any filtering, include all base modification calls in
          the summary. See filtering.md for details on filtering

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence base modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter
          threshold can be specified with by a decimal number (e.g. 0.75).
          Per-base thresholds can be specified by colon-separated values, for
          example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --include-bed &lt;INCLUDE_BED&gt;
          Only summarize base modification probabilities that are aligned to the
          positions in this BED file. (alias: include-positions)

      --mapped-only
          Only use base modification probabilities that are aligned (i.e. ignore
          soft-clipped, and inserted bases)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting
          probabilities. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or
          &lt;chrom_name&gt;
</code></pre>
<h2 id="call-mods"><a class="header" href="#call-mods">call-mods</a></h2>
<pre><code class="language-text">Call mods from a modbam, creates a new modbam with probabilities set to 100% if
a base modification is called or 0% if called canonical

Usage: modkit call-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, may be sorted and have associated index available. Can be a
          path to a file or one of `-` or `stdin` to specify a stream from
          standard input

  &lt;OUT_BAM&gt;
          Output BAM, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --ff
          Fast fail, stop processing at the first invalid sequence record.
          Default behavior is to continue and report failed/skipped records at
          the end

      --suppress-progress
          Hide the progress bar

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently
          
          [default: 4]

  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering
          threshold. If alignments are present reads will be sampled evenly
          across aligned genome. If a region is specified, either with the
          --region option or the --sample-region option, then reads will be
          sampled evenly across the region given. This option is useful for
          large BAM files. In practice, 10-50 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the
          filter-percentile. In practice, 50-100 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold. See filtering.md for details on filtering

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is
          non-deterministic, only used when no BAM index is provided

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the
          threshold probability, can be larger than the pileup processing
          interval
          
          [default: 1000000]

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global
          filter threshold can be specified with by a decimal number (e.g.
          0.75). Per-base thresholds can be specified by colon-separated values,
          for example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

      --no-filtering
          Don't filter base modification calls, assign each base modification to
          the highest probability prediction

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Filter out any base modification call that isn't part of a basecall
          sequence motif This argument can be passed multiple times. Format is
          &lt;motif_sequence&gt; &lt;offset&gt;. For example the argument to match CpG
          dinucleotides is `--motif CG 0`, or to match CG[5mC]G the argument
          would be `--motif CGCG 2`

      --cpg
          Shorthand for --motif CG 0

      --discard-motifs
          Discard base modification calls that match the provided motifs
          (instead of keeping them)

      --output-sam
          Output SAM format instead of BAM

  -h, --help
          Print help (see a summary with '-h')
</code></pre>
<h2 id="repair"><a class="header" href="#repair">repair</a></h2>
<pre><code class="language-text">Repair MM and ML tags in one bam with the correct tags from another. To use this
command, both modBAMs _must_ be sorted by read name. The "donor" modBAM's reads
must be a superset of the acceptor's reads. Extra reads in the donor are
allowed, and multiple reads with the same name (secondary, etc.) are allowed in
the acceptor. Reads with an empty SEQ field cannot be repaired and will be
rejected. Reads where there is an ambiguous alignment of the acceptor to the
donor will be rejected (and logged). See the full documentation for details

Usage: modkit repair [OPTIONS] --donor-bam &lt;DONOR_BAM&gt; --acceptor-bam &lt;ACCEPTOR_BAM&gt; --output-bam &lt;OUTPUT_BAM&gt;

Options:
  -d, --donor-bam &lt;DONOR_BAM&gt;
          Donor modBAM with original MM/ML tags. Must be sorted by read name
  -a, --acceptor-bam &lt;ACCEPTOR_BAM&gt;
          Acceptor modBAM with reads to have MM/ML base modification data
          projected on to. Must be sorted by read name
  -o, --output-bam &lt;OUTPUT_BAM&gt;
          output modBAM location
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it is recommended to use this option as some
          reads may be rejected and logged here
  -t, --threads &lt;THREADS&gt;
          The number of threads to use [default: 4]
  -h, --help
          Print help
</code></pre>
<h2 id="validate"><a class="header" href="#validate">validate</a></h2>
<pre><code class="language-text">Validate results from a set of mod-BAM files and associated BED files containing
the ground truth modified base status at reference positions

Usage: modkit validate [OPTIONS]

Options:
  -h, --help
          Print help (see a summary with '-h')

Sample Options:
      --bam-and-bed &lt;BAM&gt; &lt;BED&gt;
          Argument accepts 2 values. The first value is the BAM file path with
          modified base tags. The second is a bed file with ground truth
          reference positions. The name field in the ground truth bed file
          should be the short name (single letter code or ChEBI ID) for a
          modified base or `-` to specify a canonical base ground truth
          position. This argument can be provided more than once for multiple
          samples

  -c, --canonical-base &lt;CANONICAL_BASE&gt;
          Canonical base to evaluate. By default, this will be derived from mod
          codes in ground truth BED files. For ground truth with only canonical
          sites and/or ChEBI codes this values must be set
          
          [possible values: A, C, G, T]

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --min-identity &lt;MIN_ALIGNMENT_IDENTITY&gt;
          Only use reads with alignment identity &gt;= this number, in Q-space
          (phred score)

      --min-length &lt;MIN_ALIGNMENT_LENGTH&gt;
          Remove reads with fewer aligned reference bases than this threshold

Filtering Options:
  -p, --filter-quantile &lt;FILTER_QUANTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify modified base probability filter threshold value. If
          specified, --filter-threshold will override --filter-quantile

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

Logging Options:
      --suppress-progress
          Hide the progress bar

      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended. (alias: log)

Output Options:
  -o, --out-filepath &lt;OUT_FILEPATH&gt;
          Specify a file for machine parseable output
</code></pre>
<h2 id="pileup-hemi"><a class="header" href="#pileup-hemi">pileup-hemi</a></h2>
<pre><code class="language-text">Tabulates double-stranded base modification patters (such as hemi-methylation)
across genomic motif positions. This command produces a bedMethyl file, the
schema can be found in the online documentation

Usage: modkit pileup-hemi [OPTIONS] --ref &lt;REFERENCE_FASTA&gt; &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, should be sorted and have associated index available

Options:
  -o, --out-bed &lt;OUT_BED&gt;
          Output file to write results into. Will write to stdout if not
          provided

  -h, --help
          Print help (see a summary with '-h')

Modified Base Options:
      --cpg
          Aggregate double-stranded base modifications for CpG dinucleotides.
          This flag is short-hand for --motif CG 0

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Specify the sequence motif to pileup double-stranded base modification
          pattern counts for. The first argument should be the sequence motif
          and the second argument is the 0-based offset to the base to pileup
          base modification counts for. For example: --motif CG 0 indicates to
          generate pattern counts for the C on the top strand and the following
          C (opposite to G) on the negative strand. The motif must be
          reverse-complement palindromic or an error will be raised. See the
          documentation for more examples and details

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

      --force-allow-implicit
          Force allow implicit-canonical mode. By default modkit does not allow
          pileup with the implicit mode (e.g. C+m, no '.' or '?'). The
          `update-tags` subcommand is provided to update tags to the new mode.
          This option allows the interpretation of implicit mode tags: residues
          without modified base probability will be interpreted as being the
          non-modified base

  -k, --mask
          Respect soft masking in the reference FASTA

      --combine-mods
          Combine base modification calls, all counts of modified bases are
          summed together. See collapse.md for details

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended. (alias: log)

      --suppress-progress
          Hide the progress bar

Selection Options:
      --region &lt;REGION&gt;
          Process only the specified region of the BAM when performing pileup.
          Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;. Commas
          are allowed

      --max-depth &lt;MAX_DEPTH&gt;
          Maximum number of records to use when calculating pileup. This
          argument is passed to the pileup engine. If you have high depth data,
          consider increasing this value substantially. Must be less than
          2147483647 or an error will be raised
          
          [default: 8000]

      --include-bed &lt;INCLUDE_BED&gt;
          BED file that will restrict threshold estimation and pileup results to
          positions overlapping intervals in the file. (alias:
          include-positions)

      --include-unmapped
          Include unmapped base modifications when estimating the pass threshold

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead
          
          [default: 100000]

      --queue-size &lt;QUEUE_SIZE&gt;
          Size of queue for writing records
          
          [default: 1000]

      --chunk-size &lt;CHUNK_SIZE&gt;
          Break contigs into chunks containing this many intervals (see
          `interval_size`). This option can be used to help prevent excessive
          memory usage, usually with no performance penalty. By default, modkit
          will set this value to 1.5x the number of threads specified, so if 4
          threads are specified the chunk_size will be 6. A warning will be
          shown if this option is less than the number of threads specified

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads
          will be sampled evenly across aligned genome. If a region is
          specified, either with the --region option or the --sample-region
          option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand
          reads is sufficient to estimate the model output distribution and
          determine the filtering threshold
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the
          filter-percentile. In practice, 50-100 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold. See filtering.md for details on filtering

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is
          non-deterministic

Filtering Options:
      --no-filtering
          Do not perform any filtering, include all mod base calls in output.
          See filtering.md for details on filtering

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter
          threshold can be specified with by a decimal number (e.g. 0.75).
          Per-base thresholds can be specified by colon-separated values, for
          example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when
          estimating the threshold probability, can be larger than the pileup
          processing interval
          
          [default: 1000000]

Output Options:
      --only-tabs
          **Deprecated** The default output has all tab-delimiters. For
          bedMethyl output, separate columns with only tabs. The default is to
          use tabs for the first 10 fields and spaces thereafter. The default
          behavior is more likely to be compatible with genome viewers. Enabling
          this option may make it easier to parse the output with tabular data
          handlers that expect a single kind of separator

      --mixed-delim
          Output bedMethyl where the delimiter of columns past column 10 are
          space-delimited instead of tab-delimited. This option can be useful
          for some browsers and parsers that don't expect the extra columns of
          the bedMethyl format
</code></pre>
<h2 id="entropy"><a class="header" href="#entropy">entropy</a></h2>
<pre><code class="language-text">Use a mod-BAM to calculate methylation entropy over genomic windows

Usage: modkit entropy [OPTIONS] --in-bam &lt;IN_BAMS&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -s, --in-bam &lt;IN_BAMS&gt;
          Input mod-BAM, may be repeated multiple times to calculate entropy
          across all input mod-BAMs

  -n, --num-positions &lt;NUM_POSITIONS&gt;
          Number of modified positions to consider at a time
          
          [default: 4]

  -w, --window-size &lt;WINDOW_SIZE&gt;
          Maximum length interval that "num_positions" modified bases can occur
          in. The maximum window size decides how dense the positions are
          packed. For example, consider that the num_positions is equal to 4,
          the motif is CpG, and the window_size is equal to 8, this
          configuration would require that the modified positions are
          immediately adjacent to each other, "CGCGCGCG". On the other hand, if
          the window_size was set to 12, then multiple sequences with various
          patterns of other bases can be used CGACGATCGGCG
          
          [default: 50]

      --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format

      --mask
          Respect soft masking in the reference FASTA

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Motif to use for entropy calculation, multiple motifs can be used by
          repeating this option. When multiple motifs are used that specify
          different modified primary bases, all modification possibilities will
          be used in the calculation

      --cpg
          Use CpG motifs. Short hand for --motif CG 0 --combine-strands

      --base &lt;BASE&gt;
          Primary sequence base to calculate modification entropy on
          
          [possible values: A, C, G, T]

      --regions &lt;REGIONS_FP&gt;
          Regions over which to calculate descriptive statistics

      --combine-strands
          Combine modification counts on the positive and negative strands and
          report entropy on just the positive strand

      --min-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum coverage required at each position in the window. Windows
          without at least this many valid reads will be skipped, but positions
          within the window with enough coverage can be used by neighboring
          windows
          
          [default: 3]

      --max-filtered-positions &lt;MAX_FILTERED_POSITIONS&gt;
          Maximum number of filtered positions a read is allowed to have in a
          window, more than this number and the read will be discarded. Default
          will be 50% of `num_positions`

  -h, --help
          Print help (see a summary with '-h')

Output Options:
  -o, --out-bed &lt;OUT_BED&gt;
          Output BED file, if using `--region` this must be a directory

      --prefix &lt;PREFIX&gt;
          Only used with `--regions`, prefix files in output directory with this
          string

      --force
          Force overwrite output

      --header
          Write a header line

      --drop-zeros
          Omit windows with zero entropy

Filtering Options:
      --no-filtering
          Do not perform any filtering, include all mod base calls in output

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or for the canonical calls. When
          specified, base modification call probabilities will be required to be
          greater than or equal to this number. If `--mod-threshold` is also
          specified, _this_ value will be used for canonical calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

Sampling Options:
      --num-reads &lt;NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. Reads
          will be sampled evenly across aligned genome. If a region is
          specified, either with the --region option or the --sample-region
          option, then reads will be sampled evenly across the region given.
          This option is useful for large BAM files. In practice, 10-50 thousand
          reads is sufficient to estimate the model output distribution and
          determine the filtering threshold
          
          [default: 10042]

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of BAM-reading threads to use

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Send debug logs to this file, setting this file is recommended

      --verbose-logging
          Log regions that have zero or insufficient coverage. Requires log file

      --suppress-progress
          Hide progress bars
</code></pre>
<h2 id="localize"><a class="header" href="#localize">localize</a></h2>
<pre><code class="language-text">Investigate patterns of base modifications, by aggregating pileup counts
"localized" around genomic features of interest

Usage: modkit localize [OPTIONS] --regions &lt;REGIONS&gt; --genome-sizes &lt;GENOME_SIZES&gt; &lt;IN_BEDMETHYL&gt;

Arguments:
  &lt;IN_BEDMETHYL&gt;
          Input bedMethyl table. Should be bgzip-compressed and have an
          associated Tabix index. The tabix index will be assumed to be
          $this_file.tbi

Options:
      --regions &lt;REGIONS&gt;
          BED file of regions to calculate enrichment around. These BED records
          serve as the points from which the `--window` number of bases is
          centered

  -w, --window &lt;EXPAND_WINDOW&gt;
          Number of base pairs to search around, for example if your BED region
          records are single positions, a window of 500 will look 500 base pairs
          upstream and downstream of that position. If your region BED records
          are larger regions, this will expand from the midpoint of that region
          
          [default: 2000]

  -s, --stranded &lt;STRANDED&gt;
          Whether to only keep bedMethyl records on the "same" strand or
          "opposite" strand
          
          [possible values: same, opposite]

      --stranded-features &lt;STRANDED_FEATURES&gt;
          Force use bedMethyl records from a particular strand, default is to
          use the strand as given in the BED file (will use BOTH for BED3)
          
          [possible values: positive, negative, both]

      --min-coverage &lt;MIN_COVERAGE&gt;
          Minimum valid coverage to use a bedMethyl record
          
          [default: 3]

  -r, --genome-sizes &lt;GENOME_SIZES&gt;
          TSV of genome sizes, should be &lt;chrom&gt;\t&lt;size_in_bp&gt;

  -o, --out-file &lt;OUT_FILE&gt;
          Optionally specify a file to write output to, default is stdout

  -h, --help
          Print help (see a summary with '-h')

Output Options:
      --chart &lt;CHART_FILEPATH&gt;
          Create plots showing %-modification vs. offset. Argument should be a
          path to a file

      --name &lt;CHART_NAME&gt;
          Give the HTML document and chart a name

  -f, --force
          Force overwrite of existing output file

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file to write debug logs to

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of tabix/bgzf IO threads to use
          
          [default: 2]

      --batch-size &lt;BATCH_SIZE_BP&gt;
          [default: 500000]
</code></pre>
<h2 id="stats"><a class="header" href="#stats">stats</a></h2>
<pre><code class="language-text">Calculate base modification levels over regions

Usage: modkit stats [OPTIONS] --regions &lt;REGIONS&gt; --out-table &lt;OUT_TABLE&gt; &lt;IN_BEDMETHYL&gt;

Arguments:
  &lt;IN_BEDMETHYL&gt;  Input bedMethyl table. Should be bgzip-compressed and have an
                  associated Tabix index. The tabix index will be assumed to be
                  $this_file.tbi

Options:
      --regions &lt;REGIONS&gt;
          BED file of regions to aggregate base modification over
  -c, --mod-codes &lt;MOD_CODES&gt;
          Specify which base modification codes to use. Default will report
          information on all base modification codes encountered
  -m, --min-coverage &lt;MIN_COVERAGE&gt;
          Only use records with at least this much valid coverage [default: 1]
  -h, --help
          Print help

Output Options:
  -o, --out-table &lt;OUT_TABLE&gt;  Specify the output file to write the results
                               table, or "-"/"stdout" for stdout
      --force                  Force overwrite the output file
      --no-header              Don't add the header describing the columns to
                               the output

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;  Specify a file to write debug logs to

Compute Options:
  -t, --threads &lt;THREADS&gt;        Number of threads to use [default: 4]
      --io-threads &lt;IO_THREADS&gt;  Number of tabix/bgzf threads to use [default:
                                 2]
</code></pre>
<h2 id="open-chromatin-predict"><a class="header" href="#open-chromatin-predict">open-chromatin predict</a></h2>
<pre><code class="language-text">Usage: modkit open-chromatin predict [OPTIONS] --model &lt;MODEL_PATH&gt; --output &lt;OUTPUT&gt; &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBAM with 6mA base modification calls

Options:
      --model &lt;MODEL_PATH&gt;
          Path to directory with open-chromatin model

      --batch-size &lt;BATCH_SIZE&gt;
          Collect this many windows of data for each run through the model
          
          [default: 1024]

      --super-batch-size &lt;SUPER_BATCH_SIZE&gt;
          Number of "batches" to collect at once, see documentation for exact
          calculation or run with --dryrun to see output
          
          [default: 100]

      --step-size &lt;STEP_SIZE&gt;
          Number of base pairs to step along the genome or genomic intervals to
          make predictions. Smaller numbers will result in better resolution but
          take more computation
          
          [default: 25]

  -t, --min-coverage &lt;MIN_COVERAGE&gt;
          Require this many reads covering each prediction window
          
          [default: 5]

      --threshold &lt;FILTER_THRESHOLD&gt;
          Only emit records/windows where the open chromatin probability is
          greater than or equal to this value

  -o, --output &lt;OUTPUT&gt;
          Output bedGraph file, or "-"/"stdout" to write to stdout

      --force
          Force overwrite output file

      --log-filepath &lt;LOG_FILEPATH&gt;
          Path to file to write logs to, setting this parameter is recommended

      --include-bed &lt;INCLUDE_BED&gt;
          BED file of regions over which to make predictions

      --region &lt;REGION&gt;
          Region in the format &lt;chrom&gt;:start-end over which to make predictions

      --suppress-progress
          Hide the progress bar

      --no-header
          Don't print header in output

      --dryrun
          Perform setup and validations, but stop before performing inference

  -h, --help
          Print help (see a summary with '-h')
</code></pre>
<h2 id="extract-full"><a class="header" href="#extract-full">extract full</a></h2>
<pre><code class="language-text">Transform the probabilities from the MM/ML tags in a modBAM into a table

Usage: modkit extract full [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_PATH&gt;

Arguments:
  &lt;IN_BAM&gt;
          Path to modBAM file to extract read-level information from, or one of
          `-` or `stdin` to specify a stream from standard input. If a file is
          used it may be sorted and have associated index

  &lt;OUT_PATH&gt;
          Path to output file, "stdout" or "-" will direct output to standard
          out

Options:
      --reference &lt;REFERENCE&gt;
          Path to reference FASTA to extract reference context information from.
          Required for motif selection

  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --out-threads &lt;OUT_THREADS&gt;
          Number of threads to use for parallel bgzf writing
          
          [default: 4]

  -q, --queue-size &lt;QUEUE_SIZE&gt;
          Number of reads that can be in memory at a time. Increasing this value
          will increase thread usage, at the cost of memory usage
          
          [default: 10000]

      --ignore-index
          Ignore the BAM index (if it exists) and default to a serial scan of
          the BAM

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          Only used when an indexed modBAM is provided
          
          [default: 100000]

Output Options:
      --bgzf
          Write output as BGZF compressed file

      --force
          Force overwrite of output file

      --kmer-size &lt;KMER_SIZE&gt;
          Set the query and reference k-mer size (if a reference is provided).
          Maximum number for this value is 50
          
          [default: 5]

      --no-headers
          Don't print the header lines in the output tables

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Path to file to write run log

      --suppress-progress
          Hide the progress bar

Selection Options:
      --mapped-only
          Include only mapped bases in output (alias: mapped)

      --allow-non-primary
          Output aligned secondary and supplementary base modification
          probabilities as additional rows. The primary alignment will have all
          of the base modification probabilities (including soft-clipped ones,
          unless --mapped-only is used). The non-primary alignments will only
          have mapped bases in the output

      --num-reads &lt;NUM_READS&gt;
          Number of reads to use. Note that when using a sorted, indexed modBAM
          that the sampling algorithm will attempt to sample records evenly over
          the length of the reference sequence. The result is the final number
          of records used may be slightly more or less than the requested
          number. When piping from stdin or using a modBAM without an index, the
          requested number of reads will be the first `num_reads` records

      --region &lt;REGION&gt;
          Process only reads that are aligned to a specified region of the BAM.
          Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --include-bed &lt;INCLUDE_BED&gt;
          BED file with regions to include (alias: include-positions).
          Implicitly only includes mapped sites

  -v, --exclude-bed &lt;EXCLUDE_BED&gt;
          BED file with regions to _exclude_ (alias: exclude)

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --ignore-implicit
          Ignore implicitly canonical base modification calls. When the `.` flag
          is used in the MM tag, this implies that bases missing a base
          modification probability are to be assumed canonical. Set this flag to
          omit those base modifications from the output. For additional details
          see the SAM spec: https://samtools.github.io/hts-specs/SAMtags.pdf

Modified Base Options:
      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output read-level base modification probabilities restricted to the
          reference sequence motifs provided. The first argument should be the
          sequence motif and the second argument is the 0-based offset to the
          base to pileup base modification counts for. For example: --motif CGCG
          0 indicates include base modifications for which the read is aligned
          to the first C on the top strand and the last C (complement to G) on
          the bottom strand. The --cpg argument is short hand for --motif CG 0.
          This argument can be passed multiple times

      --annotate-motifs
          When used with `--motif` or `--cpg` emit all modified base alignment
          information even if it does not align to a reference motif, but
          annotate which aligned positions match which motifs in the "motifs"
          column. "." will be used when an aligned position does not match a
          motif

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be
          provided

  -k, --mask
          When using motifs, respect soft masking in the reference sequence

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md
</code></pre>
<h2 id="extract-calls"><a class="header" href="#extract-calls">extract calls</a></h2>
<pre><code class="language-text">Produce a table of read-level base modification calls. This table has, for each
read, one row for each base modification call in that read using the same
thresholding algorithm as in pileup, or summary (see online documentation for
details on thresholds)

Usage: modkit extract calls [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_PATH&gt;

Arguments:
  &lt;IN_BAM&gt;
          Path to modBAM file to extract read-level information from, or one of
          `-` or `stdin` to specify a stream from standard input. If a file is
          used it may be sorted and have associated index

  &lt;OUT_PATH&gt;
          Path to output file, "stdout" or "-" will direct output to standard
          out

Options:
      --reference &lt;REFERENCE&gt;
          Path to reference FASTA to extract reference context information from.
          If no reference is provided, `ref_kmer` column will be "." in the
          output. (alias: ref)

  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --out-threads &lt;OUT_THREADS&gt;
          Number of threads to use for parallel bgzf writing
          
          [default: 4]

  -q, --queue-size &lt;QUEUE_SIZE&gt;
          Number of reads that can be in memory at a time. Increasing this value
          will increase thread usage, at the cost of memory usage
          
          [default: 10000]

      --ignore-index
          Ignore the BAM index (if it exists) and default to a serial scan of
          the BAM

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          Only used when an indexed modBAM is provided
          
          [default: 100000]

Output Options:
      --bgzf
          Write output as BGZF compressed file

      --force
          Force overwrite of output file

      --kmer-size &lt;KMER_SIZE&gt;
          Set the query and reference k-mer size (if a reference is provided).
          Maximum number for this value is 50
          
          [default: 5]

      --no-headers
          Don't print the header lines in the output tables

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Path to file to write run log

      --suppress-progress
          Hide the progress bar

Selection Options:
      --mapped-only
          Include only mapped bases in output (alias: mapped)

      --allow-non-primary
          Output aligned secondary and supplementary base modification
          probabilities as additional rows. The primary alignment will have all
          of the base modification probabilities (including soft-clipped ones,
          unless --mapped-only is used). The non-primary alignments will only
          have mapped bases in the output

      --num-reads &lt;NUM_READS&gt;
          Number of reads to use. Note that when using a sorted, indexed modBAM
          that the sampling algorithm will attempt to sample records evenly over
          the length of the reference sequence. The result is the final number
          of records used may be slightly more or less than the requested
          number. When piping from stdin or using a modBAM without an index, the
          requested number of reads will be the first `num_reads` records

      --region &lt;REGION&gt;
          Process only reads that are aligned to a specified region of the BAM.
          Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --include-bed &lt;INCLUDE_BED&gt;
          BED file with regions to include (alias: include-positions).
          Implicitly only includes mapped sites

  -v, --exclude-bed &lt;EXCLUDE_BED&gt;
          BED file with regions to _exclude_ (alias: exclude)

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --ignore-implicit
          Ignore implicitly canonical base modification calls. When the `.` flag
          is used in the MM tag, this implies that bases missing a base
          modification probability are to be assumed canonical. Set this flag to
          omit those base modifications from the output. For additional details
          see the SAM spec: https://samtools.github.io/hts-specs/SAMtags.pdf

      --pass-only
          Only output base modification calls that pass the minimum confidence
          threshold. (alias: pass)

Modified Base Options:
      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Output read-level base modification probabilities restricted to the
          reference sequence motifs provided. The first argument should be the
          sequence motif and the second argument is the 0-based offset to the
          base to pileup base modification counts for. For example: --motif CGCG
          0 indicates include base modifications for which the read is aligned
          to the first C on the top strand and the last C (complement to G) on
          the bottom strand. The --cpg argument is short hand for --motif CG 0.
          This argument can be passed multiple times

      --annotate-motifs
          When used with `--motif` or `--cpg` emit all modified base alignment
          information even if it does not align to a reference motif, but
          annotate which aligned positions match which motifs in the "motifs"
          column. "." will be used when an aligned position does not match a
          motif

      --cpg
          Only output counts at CpG motifs. Requires a reference sequence to be
          provided

  -k, --mask
          When using motifs, respect soft masking in the reference sequence

      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Filtering Options:
      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter
          threshold can be specified with by a decimal number (e.g. 0.75).
          Per-base thresholds can be specified by colon-separated values, for
          example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

      --no-filtering
          Don't estimate the pass threshold, all calls will "pass"

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

Sampling Options:
      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently when
          estimating the threshold probability
          
          [default: 1000000]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the pass-threshold.
          In practice, 10-100 thousand reads is sufficient to estimate the model
          output distribution and determine the filtering threshold. See
          filtering.md for details on filtering

  -n, --sample-num-reads &lt;SAMPLE_NUM_READS&gt;
          Sample this many reads when estimating the filtering threshold. If a
          sorted, indexed modBAM is provided reads will be sampled evenly across
          aligned genome. If a region is specified, with the --region, then
          reads will be sampled evenly across the region given. This option is
          useful for large BAM files. In practice, 10-50 thousand reads is
          sufficient to estimate the model output distribution and determine the
          filtering threshold
          
          [default: 10042]

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is
          non-deterministic when using `sampling_frac`. When using `num_reads`
          the output is still deterministic
</code></pre>
<h2 id="motif-bed"><a class="header" href="#motif-bed">motif bed</a></h2>
<pre><code class="language-text">Create BED file with all locations of a sequence motif. Example: modkit motif
bed CG 0

Usage: modkit motif bed [OPTIONS] &lt;FASTA&gt; &lt;MOTIF&gt; &lt;OFFSET&gt;

Arguments:
  &lt;FASTA&gt;   Input FASTA file
  &lt;MOTIF&gt;   Motif to search for within FASTA, e.g. CG
  &lt;OFFSET&gt;  Offset within motif, e.g. 0

Options:
  -k, --mask  Respect soft masking in the reference FASTA
  -h, --help  Print help
</code></pre>
<h2 id="motif-search"><a class="header" href="#motif-search">motif search</a></h2>
<pre><code class="language-text">Search for modification-enriched subsequences in a reference genome

Usage: modkit motif search [OPTIONS] --in-bedmethyl &lt;IN_BEDMETHYL&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
      --force-override-spec
          Force override SAM specification of association of modification codes
          to primary sequence bases

  -h, --help
          Print help (see a summary with '-h')

Input Options:
  -i, --in-bedmethyl &lt;IN_BEDMETHYL&gt;
          Input bedmethyl table, can be used directly from modkit pileup

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format used for the pileup

      --contig &lt;CONTIG&gt;
          Use only bedMethyl records from this contig, requires that the
          bedMethyl be BGZIP-compressed and tabix-indexed

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of tabix/bgzf IO threads to use
          
          [default: 2]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output log to this file

      --suppress-progress
          Disable the progress bars

Search Options:
      --low-thresh &lt;LOW_THRESHOLD&gt;
          Fraction modified threshold below which consider a genome location to
          be "low modification"
          
          [default: 0.2]

      --high-thresh &lt;HIGH_THRESHOLD&gt;
          Fraction modified threshold above which consider a genome location to
          be "high modification" or enriched for modification
          
          [default: 0.6]

      --min-frac-mod &lt;FRAC_SITES_THRESH&gt;
          Minimum fraction of sites in the genome to be "high-modification" for
          a motif to be considered
          
          [default: 0.85]

      --context-size &lt;CONTEXT_SIZE&gt; &lt;CONTEXT_SIZE&gt;
          Upstream and downstream number of bases to search for a motif sequence
          around a modified base. Example: --context-size 12 12
          
          [default: 12 12]

      --min-coverage &lt;MIN_COVERAGE&gt;
          Minimum valid coverage in the bedMethyl to consider a record valid
          
          [default: 5]

      --min-sites &lt;MIN_SITES&gt;
          Minimum number of total sites in the genome required for a motif to be
          considered
          
          [default: 300]

      --min-log-odds &lt;MIN_LOG_ODDS&gt;
          Minimum log-odds to consider a motif sequence to be enriched
          
          [default: 1.5]

      --init-context-size &lt;INIT_CONTEXT_SIZE&gt; &lt;INIT_CONTEXT_SIZE&gt;
          Initial "fixed" seed window size in base pairs around the modified
          base. Example: --init-context-size 2 2
          
          [default: 2 2]

      --mod-code &lt;MOD_CODES&gt;
          Specify which modification codes to process, default will process all
          modification codes found in the input bedMethyl file

Output Options:
  -o, --out-table &lt;OUT_TABLE&gt;
          Optionally output a machine-parsable TSV (human-readable table will
          always be output to the log)

      --known-motif &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt;
          Include statistics on a suspected or known motif. Format should be
          &lt;sequence&gt; &lt;offset&gt; &lt;mod_code&gt;

      --known-motifs-table &lt;KNOWN_MOTIFS_TABLE&gt;
          Path to known motifs in tabular format. Tab-separated values:
          &lt;mod_code&gt;\t&lt;motif_seq&gt;\t&lt;offset&gt;. May have the same header as the
          output table from this command

      --eval-motifs-table &lt;OUT_KNOWN_TABLE&gt;
          Optionally output machine parsable table with known motif modification
          frequencies that were not found during search

Exhaustive Search Options:
      --exhaustive-seed-min-log-odds &lt;EXHAUSTIVE_SEED_MIN_LOG_ODDS&gt;
          Minimum log-odds to consider a motif seed sequence to be enriched when
          performing exhaustive search, decreasing this number will increase the
          number of seeds searched and thus computational time
          
          [default: 2.5]

      --exhaustive-seed-len &lt;EXHAUSTIVE_SEED_LEN&gt;
          Exhaustive search seed length, increasing this value increases
          computational time
          
          [default: 3]

      --skip-search
          Skip the exhaustive search phase, saves time but the results may be
          less sensitive

      --search-top-pct &lt;SEARCH_TOP_PCT&gt;
          During exhaustive search, instead of searching all seeds with log-odds
          above `exhaustive_seed_min_log_odds`, only search the top X-percent of
          seeds. Can be used with `min_exhaustive_seeds` and
          `max_exhaustive_seeds`

      --narrow-search
          When used in conjunction with `search_top_pct`, search the top
          X-percent of seeds, and then narrow the search space by removing
          contexts matching any motifs found. Then iterate until zero additional
          motifs are found or another stopping condition is reached

      --search-timeout &lt;SEARCH_TIMEOUT&gt;
          A stopping condition when using `--narrow-search`, stop once exaustive
          search for a modification code has been worked on for this long

      --search-batch-size &lt;SEARCH_BATCH_SIZE&gt;
          Set the batch size when performing a simple timeout on search. At
          least this many seeds will be evaluated
          
          [default: 100]

      --max-exhaustive-seeds &lt;MAX_EXHAUSTIVE_SEEDS&gt;
          Set the maximum number of exhaustive seeds to be searched in a batch.
          Overrides the X-percent of seeds to be searched when that number
          exceeds this setting

      --min-exhaustive-seeds &lt;MIN_EXHAUSTIVE_SEEDS&gt;
          Search at least this many seeds. Overrides the X-percent of seeds to
          be searched when that number is less than this setting
          
          [default: 20]

      --max-narrow-iters &lt;MAX_NARROW_ITERS&gt;
          Stopping condition when using `--narrow-search` and
          `--search-top-pct`, stop after this many iterations regardless if the
          timeout is provided and has been reached. Exaustive search will still
          stop when once no more motifs are found
</code></pre>
<h2 id="motif-evaluate"><a class="header" href="#motif-evaluate">motif evaluate</a></h2>
<pre><code class="language-text">Calculate enrichment statistics on a set of motifs from a bedMethyl table

Usage: modkit motif evaluate [OPTIONS] --in-bedmethyl &lt;IN_BEDMETHYL&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
      --force-override-spec
          Force override SAM specification of association of modification codes
          to primary sequence bases

      --suppress-table
          Don't print final table to stderr (will still go to log file)

  -h, --help
          Print help (see a summary with '-h')

Input Options:
  -i, --in-bedmethyl &lt;IN_BEDMETHYL&gt;
          Input bedmethyl table, can be used directly from modkit pileup

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format used for the pileup

      --contig &lt;CONTIG&gt;
          Use only bedMethyl records from this contig, requires that the
          bedMethyl be BGZIP-compressed and tabix-indexed

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of tabix/bgzf IO threads to use
          
          [default: 2]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output log to this file

      --suppress-progress
          Disable the progress bars

Output Options:
      --known-motif &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt;
          Format should be &lt;sequence&gt; &lt;offset&gt; &lt;mod_code&gt;

      --known-motifs-table &lt;KNOWN_MOTIFS_TABLE&gt;
          Path to known motifs in tabular format. Tab-separated values:
          &lt;mod_code&gt;\t&lt;motif_seq&gt;\t&lt;offset&gt;. May have the same header as the
          output table from this command

      --out &lt;OUT_TABLE&gt;
          Machine-parsable table of refined motifs. Human-readable table always
          printed to stderr and log

Search Options:
      --min-coverage &lt;MIN_COVERAGE&gt;
          Minimum valid coverage in the bedMethyl to consider a record valid
          
          [default: 5]

      --context-size &lt;CONTEXT_SIZE&gt; &lt;CONTEXT_SIZE&gt;
          Upstream and downstream number of bases to search for a motif sequence
          around a modified base. Example: --context-size 12 12
          
          [default: 12 12]

      --low-thresh &lt;LOW_THRESHOLD&gt;
          Fraction modified threshold below which consider a genome location to
          be "low modification"
          
          [default: 0.2]

      --high-thresh &lt;HIGH_THRESHOLD&gt;
          Fraction modified threshold above which consider a genome location to
          be "high modification" or enriched for modification
          
          [default: 0.6]
</code></pre>
<h2 id="motif-refine"><a class="header" href="#motif-refine">motif refine</a></h2>
<pre><code class="language-text">Use a previously defined list of motif sequences and further refine them with a
bedMethyl table

Usage: modkit motif refine [OPTIONS] --in-bedmethyl &lt;IN_BEDMETHYL&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
      --min_refine_frac_mod &lt;MIN_REFINE_FRAC_MODIFIED&gt;
          Minimum fraction of sites in the genome to be "high-modification" for
          a motif to be further refined, otherwise it will be discarded
          
          [default: 0.6]

      --min-refine-sites &lt;MIN_REFINE_SITES&gt;
          Minimum number of total sites in the genome required for a motif to be
          further refined, otherwise it will be discarded
          
          [default: 300]

      --force-override-spec
          Force override SAM specification of association of modification codes
          to primary sequence bases

  -h, --help
          Print help (see a summary with '-h')

Input Options:
  -i, --in-bedmethyl &lt;IN_BEDMETHYL&gt;
          Input bedmethyl table, can be used directly from modkit pileup

  -r, --ref &lt;REFERENCE_FASTA&gt;
          Reference sequence in FASTA format used for the pileup

      --contig &lt;CONTIG&gt;
          Use only bedMethyl records from this contig, requires that the
          bedMethyl be BGZIP-compressed and tabix-indexed

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of tabix/bgzf IO threads to use
          
          [default: 2]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output log to this file

      --suppress-progress
          Disable the progress bars

Output Options:
      --known-motif &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt; &lt;KNOWN_MOTIFS&gt;
          Format should be &lt;sequence&gt; &lt;offset&gt; &lt;mod_code&gt;

      --known-motifs-table &lt;KNOWN_MOTIFS_TABLE&gt;
          Path to known motifs in tabular format. Tab-separated values:
          &lt;mod_code&gt;\t&lt;motif_seq&gt;\t&lt;offset&gt;. May have the same header as the
          output table from this command

      --out &lt;OUT_TABLE&gt;
          Machine-parsable table of refined motifs. Human-readable table always
          printed to stderr and log

Search Options:
      --low-thresh &lt;LOW_THRESHOLD&gt;
          Fraction modified threshold below which consider a genome location to
          be "low modification"
          
          [default: 0.2]

      --high-thresh &lt;HIGH_THRESHOLD&gt;
          Fraction modified threshold above which consider a genome location to
          be "high modification" or enriched for modification
          
          [default: 0.6]

      --min-frac-mod &lt;FRAC_SITES_THRESH&gt;
          Minimum fraction of sites in the genome to be "high-modification" for
          a motif to be considered
          
          [default: 0.85]

      --context-size &lt;CONTEXT_SIZE&gt; &lt;CONTEXT_SIZE&gt;
          Upstream and downstream number of bases to search for a motif sequence
          around a modified base. Example: --context-size 12 12
          
          [default: 12 12]

      --min-coverage &lt;MIN_COVERAGE&gt;
          Minimum valid coverage in the bedMethyl to consider a record valid
          
          [default: 5]

      --min-sites &lt;MIN_SITES&gt;
          Minimum number of total sites in the genome required for a motif to be
          considered
          
          [default: 300]

      --min-log-odds &lt;MIN_LOG_ODDS&gt;
          Minimum log-odds to consider a motif sequence to be enriched
          
          [default: 1.5]
</code></pre>
<h2 id="dmr-pair"><a class="header" href="#dmr-pair">dmr pair</a></h2>
<pre><code class="language-text">Compare regions in a pair of samples (for example, tumor and normal or control
and experiment). A sample is input as a bgzip pileup bedMethyl (produced by
pileup, for example) that has an associated tabix index. Output is a BED file
with the score column indicating the magnitude of the difference in methylation
between the two samples. See the online documentation for additional details

Usage: modkit dmr pair [OPTIONS] --ref &lt;REFERENCE_FASTA&gt;

Options:
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should
          be tab-separated (spaces allowed in the "name" column). Requires
          chrom, chromStart and chromEnd. The Name column is optional. Strand is
          currently ignored. When omitted, methylation levels are compared at
          each site

      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for used in the pileup/alignment

  -h, --help
          Print help (see a summary with '-h')

Sample Options:
  -a &lt;CONTROL_BED_METHYL&gt;
          Bgzipped bedMethyl file for the first (usually control) sample. There
          should be a tabix index with the same name and .tbi next to this file
          or the --index-a option must be provided

  -b &lt;EXP_BED_METHYL&gt;
          Bgzipped bedMethyl file for the second (usually experimental) sample.
          There should be a tabix index with the same name and .tbi next to this
          file or the --index-b option must be provided

  -m, --base &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to
          calculate differentially methylated regions using only cytosine
          modifications use --base C

      --assign-code &lt;MOD_CODE_ASSIGNMENTS&gt;
          Extra assignments of modification codes to their respective primary
          bases. In general, modkit dmr will use the SAM specification to know
          which modification codes are appropriate to use for a given primary
          base. For example "h" is the code for 5hmC, so is appropriate for
          cytosine bases, but not adenine bases. However, if your bedMethyl file
          contains custom codes or codes that are not part of the specification,
          you can specify which primary base they belong to here with
          --assign-code x:C meaning associate modification code "x" with
          cytosine (C) primary sequence bases. If a code is encountered that is
          not part of the specification, the bedMethyl record will not be used,
          this will be logged

  -k, --mask
          Respect soft masking in the reference FASTA

      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See
          the help for pileup for the specification and description of valid
          coverage
          
          [default: 0]

Output Options:
  -o, --out-path &lt;OUT_PATH&gt;
          Path to file to direct output, optional, no argument will direct
          output to stdout

      --header
          Include header in output

Segmentation Options:
      --segment &lt;SEGMENTATION_FP&gt;
          Run segmentation, output segmented differentially methylated regions
          to this file

      --max-gap-size &lt;MAX_GAP_SIZE&gt;
          Maximum number of base pairs between modified bases for them to be
          segmented together
          
          [default: 5000]

      --dmr-prior &lt;DMR_PRIOR&gt;
          Prior probability of a differentially methylated position
          
          [default: 0.1]

      --diff-stay &lt;DIFF_STAY&gt;
          Maximum probability of continuing a differentially methylated block,
          decay will be dynamic based on proximity to the next position
          
          [default: 0.9]

      --significance-factor &lt;SIGNIFICANCE_FACTOR&gt;
          Significance factor, effective p-value necessary to favor the
          "Different" state
          
          [default: 0.01]

      --log-transition-decay
          Use logarithmic decay for "Different" stay probability

      --decay-distance &lt;DECAY_DISTANCE&gt;
          After this many base pairs, the transition probability will become the
          prior probability of encountering a differentially modified position
          
          [default: 500]

      --fine-grained
          Preset HMM segmentation parameters for higher propensity to switch
          from "Same" to "Different" state. Results will be shorter segments,
          but potentially higher sensitivity

Logging Options:
      --careful
          Log out which sequences are in common between the samples and the
          reference FASTA, useful for debugging

      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option

      --suppress-progress
          Don't show progress bars

      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt;
          ignore regions that are not found in the tabix header warn =&gt; log
          (debug) regions that are missing fatal =&gt; log (error) and exit the
          program when a region is missing
          
          [default: quiet]
          [possible values: quiet, warn, fail]

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --io-threads &lt;IO_THREADS&gt;
          Number of threads to use when for decompression
          
          [default: 4]

      --batch-size &lt;BATCH_SIZE&gt;
          Control the  batch size. The batch size is the number of regions to
          load at a time. Each region will be processed concurrently. Loading
          more regions at a time will decrease IO to load data, but will use
          more memory. Default will be 50% more than the number of threads
          assigned

  -f, --force
          Force overwrite of output file, if it already exists

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead
          
          [default: 100000]

Single-site Options:
      --prior &lt;PRIOR&gt; &lt;PRIOR&gt;
          Prior distribution for estimating MAP-based p-value. Should be two
          arguments for alpha and beta (e.g. 1.0 1.0). See
          `dmr_scoring_details.md` for additional details on how the metric is
          calculated

      --delta &lt;DELTA&gt;
          Consider only effect sizes greater than this when calculating the
          MAP-based p-value
          
          [default: 0.05]

  -N, --n-sample-records &lt;N_SAMPLE_RECORDS&gt;
          Sample this many reads when estimating the max coverage thresholds
          
          [default: 10042]

      --max-coverages &lt;MAX_COVERAGES&gt; &lt;MAX_COVERAGES&gt;
          Max coverages to enforce when calculating estimated MAP-based p-value

      --cap-coverages
          When using replicates, cap coverage to be equal to the maximum
          coverage for a single sample. For example, if there are 3 replicates
          with max_coverage of 30, the total coverage would normally be 90.
          Using --cap-coverages will down sample the data to 30X
</code></pre>
<h2 id="dmr-multi"><a class="header" href="#dmr-multi">dmr multi</a></h2>
<pre><code class="language-text">Compare regions between all pairs of samples (for example a trio sample set or
haplotyped trio sample set). As with `pair` all inputs must be bgzip compressed
bedMethyl files with associated tabix indices. Each sample must be assigned a
name. Output is a directory of BED files with the score column indicating the
magnitude of the difference in methylation between the two samples indicated in
the file name. See the online documentation for additional details

Usage: modkit dmr multi [OPTIONS] --regions-bed &lt;REGIONS_BED&gt; --out-dir &lt;OUT_DIR&gt; --ref &lt;REFERENCE_FASTA&gt;

Options:
  -h, --help  Print help

Sample Options:
  -s, --sample &lt;SAMPLES&gt; &lt;SAMPLES&gt;
          Two or more named samples to compare. Two arguments are required
          &lt;path&gt; &lt;name&gt;. This option should be repeated at least two times. When
          two samples have the same name, they will be combined
  -r, --regions-bed &lt;REGIONS_BED&gt;
          BED file of regions over which to compare methylation levels. Should
          be tab-separated (spaces allowed in the "name" column). Requires
          chrom, chromStart and chromEnd. The Name column is optional. Strand is
          currently ignored
      --ref &lt;REFERENCE_FASTA&gt;
          Path to reference fasta for the pileup
  -m, --base &lt;MODIFIED_BASES&gt;
          Bases to use to calculate DMR, may be multiple. For example, to
          calculate differentially methylated regions using only cytosine
          modifications use --base C
      --assign-code &lt;MOD_CODE_ASSIGNMENTS&gt;
          Extra assignments of modification codes to their respective primary
          bases. In general, modkit dmr will use the SAM specification to know
          which modification codes are appropriate to use for a given primary
          base. For example "h" is the code for 5hmC, so is appropriate for
          cytosine bases, but not adenine bases. However, if your bedMethyl file
          contains custom codes or codes that are not part of the specification,
          you can specify which primary base they belong to here with
          --assign-code x:C meaning associate modification code "x" with
          cytosine (C) primary sequence bases. If a code is encountered that is
          not part of the specification, the bedMethyl record will not be used,
          this will be logged
  -k, --mask
          Respect soft masking in the reference FASTA
      --min-valid-coverage &lt;MIN_VALID_COVERAGE&gt;
          Minimum valid coverage required to use an entry from a bedMethyl. See
          the help for pileup for the specification and description of valid
          coverage [default: 0]

Output Options:
      --header             Include header in output
  -o, --out-dir &lt;OUT_DIR&gt;  Directory to place output DMR results in BED format
  -p, --prefix &lt;PREFIX&gt;    Prefix files in directory with this label
  -f, --force              Force overwrite of output file, if it already exists

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it's recommended to use this option
      --suppress-progress
          Don't show progress bars
      --missing &lt;HANDLE_MISSING&gt;
          How to handle regions found in the `--regions` BED file. quiet =&gt;
          ignore regions that are not found in the tabix header warn =&gt; log
          (debug) regions that are missing fatal =&gt; log (error) and exit the
          program when a region is missing [default: quiet] [possible values:
          quiet, warn, fail]

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use [default: 4]
      --io-threads &lt;IO_THREADS&gt;
          Number of threads to use when for decompression [default: 4]
</code></pre>
<h2 id="bedmethyl-merge"><a class="header" href="#bedmethyl-merge">bedmethyl merge</a></h2>
<pre><code class="language-text">Perform an outer join on two or more bedMethyl files, summing their counts for
records that overlap

Usage: modkit bedmethyl merge [OPTIONS] --out-bed &lt;OUT_BED&gt; --genome-sizes &lt;GENOME_SIZES&gt; [IN_BEDMETHYL] [IN_BEDMETHYL]...

Arguments:
  [IN_BEDMETHYL] [IN_BEDMETHYL]...
          Input bedMethyl table(s). Should be bgzip-compressed and have an
          associated Tabix index. The tabix index will be assumed to be
          $this_file.tbi

Options:
  -g, --genome-sizes &lt;GENOME_SIZES&gt;
          TSV of genome sizes, should be &lt;chrom&gt;\t&lt;size_in_bp&gt;

  -h, --help
          Print help (see a summary with '-h')

Output Options:
  -o, --out-bed &lt;OUT_BED&gt;
          Specify the output file to write the results table

      --force
          Force overwrite the output file

      --header
          Output a header with the bedMethyl

      --mixed-delim
          Output bedMethyl where the delimiter of columns past column 10 are
          space-delimited instead of tab-delimited. This option can be useful
          for some browsers and parsers that don't expect the extra columns of
          the bedMethyl format

Compute Options:
      --chunk-size &lt;CHUNK_SIZE&gt;
          Chunk size for how many start..end regions for each chromosome to
          read. Larger values will lead to faster merging at the expense of
          memory usage, while smaller values will be slower with lower memory
          usage. This option will only impact large bedmethyl files

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead
          
          [default: 100000]

  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --queue-size &lt;QUEUE_SIZE&gt;
          Number of batches (of size chunk size) allowed to be in a pre-written
          state at once. Increasing this number will increase memory usage
          
          [default: 30]

      --io-threads &lt;IO_THREADS&gt;
          Number of tabix/bgzf threads to use
          
          [default: 2]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file to write debug logs to
</code></pre>
<h2 id="bedmethyl-tobigwig"><a class="header" href="#bedmethyl-tobigwig">bedmethyl tobigwig</a></h2>
<pre><code class="language-text">Make a BigWig track from a bedMethyl file or stream. For details on the BigWig
format see https://doi.org/10.1093/bioinformatics/btq351

Usage: modkit bedmethyl tobigwig [OPTIONS] --mod-codes &lt;MOD_CODES&gt; &lt;--sizes &lt;CHROMSIZES&gt;|--header &lt;INPUT_BAM&gt;&gt; &lt;IN_BEDMETHYL&gt; &lt;OUT_FP&gt;

Arguments:
  &lt;IN_BEDMETHYL&gt;
          Input bedmethyl, uncompressed, "-" or "stdin" indicates an input
          stream

  &lt;OUT_FP&gt;
          Output bigWig filename

Options:
  -g, --sizes &lt;CHROMSIZES&gt;
          A chromosome sizes file. Each line should be a chromosome and its size
          in bases, separated by whitespace. A fasta index (.fai) works as well.
          Use instead of the bam header

  -b, --header &lt;INPUT_BAM&gt;
          modBAM from which the pileup was generated. Chromosome sizes are
          gathered from the header. Use instead of the chromosome sizes file

  -m, --mod-codes &lt;MOD_CODES&gt;
          Make a bigWig track where the values are the percent of bases with
          this modification, use multiple comma-separated codes to combine
          counts. For example --mod-code m makes a track of the 5mC percentages
          and --mod-codes h,m will make a track of the combined counts from 5hmC
          and 5mC. Combining counts for different primary bases will cause an
          error (e.g. --mod-codes a,h will error)

  -h, --help
          Print help (see a summary with '-h')

Output Options:
      --negative-strand-values
          Report the percentages on the negative strand as negative values. The
          data range will be [-100, 100]

  -z, --nzooms &lt;NZOOMS&gt;
          Set the maximum of zooms to create
          
          [default: 10]

      --zooms &lt;ZOOMS&gt;...
          Set the zoom resolutions to use (overrides the --nzooms argument)

  -u, --uncompressed
          Don't use compression

      --block-size &lt;BLOCK_SIZE&gt;
          Number of items to bundle in r-tree
          
          [default: 256]

      --items-per-slot &lt;ITEMS_PER_SLOT&gt;
          Number of data points bundled at lowest level
          
          [default: 1024]

Compute Options:
  -t, --nthreads &lt;NTHREADS&gt;
          Set the number of threads to use. This tool will typically use ~225%
          CPU on a HDD. SDDs may be higher. (IO bound)
          
          [default: 6]

      --inmemory
          Do not create temporary files for intermediate data

      --force-chromosome-ordering
          If input bedMethyl has sorting of the same scheme as `sort`, this
          option may speed up conversion

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended. (alias: log)

      --suppress-progress
          Hide the progress bar
</code></pre>
<h2 id="modbam-adjust-mods"><a class="header" href="#modbam-adjust-mods">modbam adjust-mods</a></h2>
<pre><code class="language-text">Performs various operations on BAM files containing base modification
information, such as converting base modification codes and ignoring
modification calls. Produces a BAM output file

Usage: modkit modbam adjust-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM file, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

  &lt;OUT_BAM&gt;
          File path to new BAM file to be created. Can be a path to a file or
          one of `-` or `stdin` to specify a stream from standard output

Options:
  -f, --ff
          Fast fail, stop processing at the first invalid sequence record.
          Default behavior is to continue and report failed/skipped records at
          the end

  -h, --help
          Print help (see a summary with '-h')

Output Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Output debug logs to file at this path

      --output-sam
          Output SAM format instead of BAM

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Modified base code to ignore/remove, see
          https://samtools.github.io/hts-specs/SAMtags.pdf for details on the
          modified base codes

      --convert &lt;CONVERT&gt; &lt;CONVERT&gt;
          Convert one mod-tag to another, summing the probabilities together if
          the retained mod tag is already present

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Filter out any base modification call that isn't part of a basecall
          sequence motif. This argument can be passed multiple times. Format is
          &lt;motif_sequence&gt; &lt;offset&gt;. For example the argument to match CpG
          dinucleotides is `--motif CG 0`, or to match CG[5mC]G the argument
          would be `--motif CGCG 2`. Single bases can be used as motifs to keep
          only base modification calls for a specific primary base, for example
          `--motif C 0`

      --cpg
          Shorthand for --motif CG 0

      --discard-motifs
          Discard base modification calls that match the provided motifs
          (instead of keeping them)

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --filter-probs
          Filter out the lowest confidence base modification probabilities

      --mapped-only
          Only use base modification probabilities from bases that are aligned
          when estimating the filter threshold (i.e. ignore soft-clipped, and
          inserted bases)

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering
          threshold. If alignments are present reads will be sampled evenly
          across aligned genome. If a region is specified, either with the
          --region option or the --sample-region option, then reads will be
          sampled evenly across the region given. This option is useful for
          large BAM files. In practice, 10-50 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold
          
          [default: 10042]

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the
          threshold probability, can be larger than the pileup processing
          interval
          
          [default: 1000000]

Filtering Options:
  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global
          filter threshold can be specified with by a decimal number (e.g.
          0.75). Per-base thresholds can be specified by colon-separated values,
          for example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

Logging Options:
      --suppress-progress
          Hide the progress bar
</code></pre>
<h2 id="modbam-update-tags"><a class="header" href="#modbam-update-tags">modbam update-tags</a></h2>
<pre><code class="language-text">Renames Mm/Ml to tags to MM/ML. Also allows changing the mode flag from silent
'.' to explicitly '?' or '.'

Usage: modkit modbam update-tags [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;   BAM to update modified base tags in. Can be a path to a file or one
             of `-` or `stdin` to specify a stream from standard input
  &lt;OUT_BAM&gt;  File to new BAM file to be created or one of `-` or `stdin` to
             specify a stream from standard output

Options:
  -m, --mode &lt;MODE&gt;        Mode, change mode to this value, options {'explicit',
                           'implicit'}. See spec at:
                           https://samtools.github.io/hts-specs/SAMtags.pdf.
                           'explicit' ('?') means residues without modification
                           probabilities will not be assumed canonical or
                           modified. 'implicit' means residues without explicit
                           modification probabilities are assumed to be
                           canonical [possible values: explicit, implicit]
      --no-implicit-probs  Don't add implicit canonical calls. This flag is
                           important when converting from one of the implicit
                           modes ( `.` or `""`) to explicit mode (`?`). By
                           passing this flag, the bases without associated base
                           modification probabilities will not be assumed to be
                           canonical. No base modification probability will be
                           written for these bases, meaning there is no
                           information. The mode will automatically be set to
                           the explicit mode `?`
  -h, --help               Print help

Compute Options:
  -t, --threads &lt;THREADS&gt;  Number of threads to use [default: 4]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;  Output debug logs to file at this path

Output Options:
      --output-sam  Output SAM format instead of BAM
</code></pre>
<h2 id="modbam-sample-probs"><a class="header" href="#modbam-sample-probs">modbam sample-probs</a></h2>
<pre><code class="language-text">Calculate an estimate of the base modification probability distribution

Usage: modkit modbam sample-probs [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM with modified base tags. If a index is found reads will be
          sampled evenly across the length of the reference sequence. Can be a
          path to a file or one of `-` or `stdin` to specify a stream from
          standard input

Options:
  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          Interval chunk size in base pairs to process concurrently. Smaller
          interval chunk sizes will use less memory but incur more overhead.
          Only used when sampling probs from an indexed bam
          
          [default: 1000000]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --suppress-progress
          Hide the progress bar

Output Options:
  -p, --percentiles &lt;PERCENTILES&gt;
          Percentiles to calculate, a space separated list of floats
          
          [default: 0.1,0.5,0.9]

  -o, --out-dir &lt;OUT_DIR&gt;
          Directory to deposit result tables into. Required for model
          probability histogram output

      --prefix &lt;PREFIX&gt;
          Label to prefix output files with

      --force
          Overwrite results if present

      --hist
          Output histogram of base modification prediction probabilities

      --dna-color &lt;PRIMARY_BASE_COLORS&gt; &lt;PRIMARY_BASE_COLORS&gt;
          Set colors of primary bases in histogram, should be RGB format, e.g.
          "#0000FF" is defailt for canonical cytosine

      --mod-color &lt;MOD_BASE_COLORS&gt; &lt;MOD_BASE_COLORS&gt;
          Set colors of modified bases in histogram, should be RGB format, e.g.
          "#FF00FF" is default for 5hmC

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting
          probabilities. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or
          &lt;chrom_name&gt;

      --include-bed &lt;INCLUDE_BED&gt;
          Only sample base modification probabilities that are aligned to the
          positions in this BED file. (alias: include-positions)

      --mapped-only
          Only use base modification probabilities that are aligned (i.e. ignore
          soft-clipped, and inserted bases)

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended
          when using a large BAM without an index. If an indexed BAM is
          provided, the reads will be sampled evenly over the length of the
          aligned reference. If a region is passed with the --region option,
          they will be sampled over the genomic region. Actual number of reads
          used may deviate slightly from this number
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of
          reads to sample, for example 0.1 will sample 1/10th of the reads

      --no-sampling
          No sampling, use all of the reads to calculate the filter thresholds

  -s, --seed &lt;SEED&gt;
          Random seed for deterministic running, the default is
          non-deterministic, only used when no BAM index is provided
</code></pre>
<h2 id="modbam-summary"><a class="header" href="#modbam-summary">modbam summary</a></h2>
<pre><code class="language-text">Summarize the mod tags present in a BAM and get basic statistics. The default
output is a totals table (designated by '#' lines) and a modification calls
table. Descriptions of the columns can be found in the README

Usage: modkit modbam summary [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBam, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

Options:
  -h, --help
          Print help (see a summary with '-h')

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          When using regions, interval chunk size in base pairs to process
          concurrently. Smaller interval chunk sizes will use less memory but
          incur more overhead
          
          [default: 1000000]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --suppress-progress
          Hide the progress bar

Output Options:
      --tsv
          Output summary as a tab-separated variables stdout instead of a table

Sampling Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended
          when using a large BAM without an index. If an indexed BAM is
          provided, the reads will be sampled evenly over the length of the
          aligned reference. If a region is passed with the --region option,
          they will be sampled over the genomic region. Actual number of reads
          used may deviate slightly from this number
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Instead of using a defined number of reads, specify a fraction of
          reads to sample when estimating the filter threshold. For example 0.1
          will sample 1/10th of the reads

      --no-sampling
          No sampling, use all the reads to calculate the filter thresholds and
          generating the summary

  -s, --seed &lt;SEED&gt;
          Sets a random seed for deterministic running (when using
          --sample-frac), the default is non-deterministic, only used when no
          BAM index is provided

Filtering Options:
      --no-filtering
          Do not perform any filtering, include all base modification calls in
          the summary. See filtering.md for details on filtering

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence base modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per-base. Global filter
          threshold can be specified with by a decimal number (e.g. 0.75).
          Per-base thresholds can be specified by colon-separated values, for
          example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

Modified Base Options:
      --ignore &lt;IGNORE&gt;
          Ignore a modified base class  _in_situ_ by redistributing base
          modification probability equally across other options. For example, if
          collapsing 'h', with 'm' and canonical options, half of the
          probability of 'h' will be added to both 'm' and 'C'. A full
          description of the methods can be found in collapse.md

Selection Options:
      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --include-bed &lt;INCLUDE_BED&gt;
          Only summarize base modification probabilities that are aligned to the
          positions in this BED file. (alias: include-positions)

      --mapped-only
          Only use base modification probabilities that are aligned (i.e. ignore
          soft-clipped, and inserted bases)

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting
          probabilities. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or
          &lt;chrom_name&gt;
</code></pre>
<h2 id="modbam-call-mods"><a class="header" href="#modbam-call-mods">modbam call-mods</a></h2>
<pre><code class="language-text">Call mods from a modbam, creates a new modbam with probabilities set to 100% if
a base modification is called or 0% if called canonical

Usage: modkit modbam call-mods [OPTIONS] &lt;IN_BAM&gt; &lt;OUT_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input BAM, may be sorted and have associated index available. Can be a
          path to a file or one of `-` or `stdin` to specify a stream from
          standard input

  &lt;OUT_BAM&gt;
          Output BAM, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --ff
          Fast fail, stop processing at the first invalid sequence record.
          Default behavior is to continue and report failed/skipped records at
          the end

      --suppress-progress
          Hide the progress bar

  -t, --threads &lt;THREADS&gt;
          Number of threads to use while processing chunks concurrently
          
          [default: 4]

  -n, --num-reads &lt;NUM_READS&gt;
          Sample approximately this many reads when estimating the filtering
          threshold. If alignments are present reads will be sampled evenly
          across aligned genome. If a region is specified, either with the
          --region option or the --sample-region option, then reads will be
          sampled evenly across the region given. This option is useful for
          large BAM files. In practice, 10-50 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold
          
          [default: 10042]

  -f, --sampling-frac &lt;SAMPLING_FRAC&gt;
          Sample this fraction of the reads when estimating the
          filter-percentile. In practice, 50-100 thousand reads is sufficient to
          estimate the model output distribution and determine the filtering
          threshold. See filtering.md for details on filtering

      --seed &lt;SEED&gt;
          Set a random seed for deterministic running, the default is
          non-deterministic, only used when no BAM index is provided

      --sample-region &lt;SAMPLE_REGION&gt;
          Specify a region for sampling reads from when estimating the threshold
          probability. If this option is not provided, but --region is provided,
          the genomic interval passed to --region will be used. Format should be
          &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or &lt;chrom_name&gt;

      --sampling-interval-size &lt;SAMPLING_INTERVAL_SIZE&gt;
          Interval chunk size to process concurrently when estimating the
          threshold probability, can be larger than the pileup processing
          interval
          
          [default: 1000000]

  -p, --filter-percentile &lt;FILTER_PERCENTILE&gt;
          Filter out modified base calls where the probability of the predicted
          variant is below this confidence percentile. For example, 0.1 will
          filter out the 10% lowest confidence modification calls
          
          [default: 0.1]

      --filter-threshold &lt;FILTER_THRESHOLD&gt;
          Specify the filter threshold globally or per primary base. A global
          filter threshold can be specified with by a decimal number (e.g.
          0.75). Per-base thresholds can be specified by colon-separated values,
          for example C:0.75 specifies a threshold value of 0.75 for cytosine
          modification calls. Additional per-base thresholds can be specified by
          repeating the option: for example --filter-threshold C:0.75
          --filter-threshold A:0.70 or specify a single base option and a
          default for all other bases with: --filter-threshold A:0.70
          --filter-threshold 0.9 will specify a threshold value of 0.70 for
          adenine and 0.9 for all other base modification calls

      --mod-threshold &lt;MOD_THRESHOLDS&gt;
          Specify a passing threshold to use for a base modification,
          independent of the threshold for the primary sequence base or the
          default. For example, to set the pass threshold for 5hmC to 0.8 use
          `--mod-threshold h:0.8`. The pass threshold will still be estimated as
          usual and used for canonical cytosine and other modifications unless
          the `--filter-threshold` option is also passed. See the online
          documentation for more details

      --no-filtering
          Don't filter base modification calls, assign each base modification to
          the highest probability prediction

      --edge-filter &lt;EDGE_FILTER&gt;
          Discard base modification calls that are this many bases from the
          start or the end of the read. Two comma-separated values may be
          provided to asymmetrically filter out base modification calls from the
          start and end of the reads. For example, 4,8 will filter out base
          modification calls in the first 4 and last 8 bases of the read

      --invert-edge-filter
          Invert the edge filter, instead of filtering out base modification
          calls at the ends of reads, only _keep_ base modification calls at the
          ends of reads. E.g. if usually, "4,8" would remove (i.e. filter out)
          base modification calls in the first 4 and last 8 bases of the read,
          using this flag will keep only base modification calls in the first 4
          and last 8 bases

      --motif &lt;MOTIF&gt; &lt;MOTIF&gt;
          Filter out any base modification call that isn't part of a basecall
          sequence motif This argument can be passed multiple times. Format is
          &lt;motif_sequence&gt; &lt;offset&gt;. For example the argument to match CpG
          dinucleotides is `--motif CG 0`, or to match CG[5mC]G the argument
          would be `--motif CGCG 2`

      --cpg
          Shorthand for --motif CG 0

      --discard-motifs
          Discard base modification calls that match the provided motifs
          (instead of keeping them)

      --output-sam
          Output SAM format instead of BAM

  -h, --help
          Print help (see a summary with '-h')
</code></pre>
<h2 id="modbam-repair"><a class="header" href="#modbam-repair">modbam repair</a></h2>
<pre><code class="language-text">Repair MM and ML tags in one bam with the correct tags from another. To use this
command, both modBAMs _must_ be sorted by read name. The "donor" modBAM's reads
must be a superset of the acceptor's reads. Extra reads in the donor are
allowed, and multiple reads with the same name (secondary, etc.) are allowed in
the acceptor. Reads with an empty SEQ field cannot be repaired and will be
rejected. Reads where there is an ambiguous alignment of the acceptor to the
donor will be rejected (and logged). See the full documentation for details

Usage: modkit modbam repair [OPTIONS] --donor-bam &lt;DONOR_BAM&gt; --acceptor-bam &lt;ACCEPTOR_BAM&gt; --output-bam &lt;OUTPUT_BAM&gt;

Options:
  -d, --donor-bam &lt;DONOR_BAM&gt;
          Donor modBAM with original MM/ML tags. Must be sorted by read name
  -a, --acceptor-bam &lt;ACCEPTOR_BAM&gt;
          Acceptor modBAM with reads to have MM/ML base modification data
          projected on to. Must be sorted by read name
  -o, --output-bam &lt;OUTPUT_BAM&gt;
          output modBAM location
      --log-filepath &lt;LOG_FILEPATH&gt;
          File to write logs to, it is recommended to use this option as some
          reads may be rejected and logged here
  -t, --threads &lt;THREADS&gt;
          The number of threads to use [default: 4]
  -h, --help
          Print help
</code></pre>
<h2 id="modbam-check-tags"><a class="header" href="#modbam-check-tags">modbam check-tags</a></h2>
<pre><code class="language-text">Usage: modkit modbam check-tags [OPTIONS] &lt;IN_BAM&gt;

Arguments:
  &lt;IN_BAM&gt;
          Input modBam, can be a path to a file or one of `-` or `stdin` to
          specify a stream from standard input

Options:
      --permissive
          Don't exit 1 when invalid records are found in the input

  -h, --help
          Print help (see a summary with '-h')

IO Options:
  -o, --out-dir &lt;OUT_DIR&gt;
          Write output tables into this directory. The directory will be created
          if it doesn't exist

  -f, --force
          Force overwrite of previous output

      --prefix &lt;PREFIX&gt;
          Prefix output files with this string

Compute Options:
  -t, --threads &lt;THREADS&gt;
          Number of threads to use
          
          [default: 4]

      --ignore-index
          Perform a linear scan of the modBAM even if the index is found

  -i, --interval-size &lt;INTERVAL_SIZE&gt;
          When using regions, interval chunk size in base pairs to process
          concurrently. Smaller interval chunk sizes will use less memory but
          incur more overhead
          
          [default: 5000000]

Logging Options:
      --log-filepath &lt;LOG_FILEPATH&gt;
          Specify a file for debug logs to be written to, otherwise ignore them.
          Setting a file is recommended

      --suppress-progress
          Hide the progress bar

Selection Options:
  -n, --num-reads &lt;NUM_READS&gt;
          Approximate maximum number of reads to use, especially recommended
          when using a large BAM without an index. If an indexed BAM is
          provided, the reads will be sampled evenly over the length of the
          aligned reference. If a region is passed with the --region option,
          they will be sampled over the genomic region. Actual number of reads
          used may deviate slightly from this number

      --allow-non-primary
          Check tags on non-primary alignments as well. Keep in mind this may
          incur a double-counting of the read with its primary mapping

      --mapped-only
          Only check alignments that are mapped

      --region &lt;REGION&gt;
          Process only the specified region of the BAM when collecting
          probabilities. Format should be &lt;chrom_name&gt;:&lt;start&gt;-&lt;end&gt; or
          &lt;chrom_name&gt;

      --head &lt;HEAD&gt;
          Use the first N reads without sampling. Shorthand for --ignore-index
          --num-reads N
</code></pre>

                    </main>

                    <nav class="nav-wrapper" aria-label="Page navigation">
                        <!-- Mobile navigation buttons -->
                            <a rel="prev" href="intro_open_chromatin.html" class="mobile-nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                                <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"/></svg></span>
                            </a>

                            <a rel="next prefetch" href="troubleshooting.html" class="mobile-nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                                <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"/></svg></span>
                            </a>

                        <div style="clear: both"></div>
                    </nav>
                </div>
            </div>

            <nav class="nav-wide-wrapper" aria-label="Page navigation">
                    <a rel="prev" href="intro_open_chromatin.html" class="nav-chapters previous" title="Previous chapter" aria-label="Previous chapter" aria-keyshortcuts="Left">
                        <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M41.4 233.4c-12.5 12.5-12.5 32.8 0 45.3l160 160c12.5 12.5 32.8 12.5 45.3 0s12.5-32.8 0-45.3L109.3 256 246.6 118.6c12.5-12.5 12.5-32.8 0-45.3s-32.8-12.5-45.3 0l-160 160z"/></svg></span>
                    </a>

                    <a rel="next prefetch" href="troubleshooting.html" class="nav-chapters next" title="Next chapter" aria-label="Next chapter" aria-keyshortcuts="Right">
                        <span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 320 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M278.6 233.4c12.5 12.5 12.5 32.8 0 45.3l-160 160c-12.5 12.5-32.8 12.5-45.3 0s-12.5-32.8 0-45.3L210.7 256 73.4 118.6c-12.5-12.5-12.5-32.8 0-45.3s32.8-12.5 45.3 0l160 160z"/></svg></span>
                    </a>
            </nav>

        </div>

        <template id=fa-eye><span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 576 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M288 32c-80.8 0-145.5 36.8-192.6 80.6C48.6 156 17.3 208 2.5 243.7c-3.3 7.9-3.3 16.7 0 24.6C17.3 304 48.6 356 95.4 399.4C142.5 443.2 207.2 480 288 480s145.5-36.8 192.6-80.6c46.8-43.5 78.1-95.4 93-131.1c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C433.5 68.8 368.8 32 288 32zM432 256c0 79.5-64.5 144-144 144s-144-64.5-144-144s64.5-144 144-144s144 64.5 144 144zM288 192c0 35.3-28.7 64-64 64c-11.5 0-22.3-3-31.6-8.4c-.2 2.8-.4 5.5-.4 8.4c0 53 43 96 96 96s96-43 96-96s-43-96-96-96c-2.8 0-5.6 .1-8.4 .4c5.3 9.3 8.4 20.1 8.4 31.6z"/></svg></span></template>
        <template id=fa-eye-slash><span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 640 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M38.8 5.1C28.4-3.1 13.3-1.2 5.1 9.2S-1.2 34.7 9.2 42.9l592 464c10.4 8.2 25.5 6.3 33.7-4.1s6.3-25.5-4.1-33.7L525.6 386.7c39.6-40.6 66.4-86.1 79.9-118.4c3.3-7.9 3.3-16.7 0-24.6c-14.9-35.7-46.2-87.7-93-131.1C465.5 68.8 400.8 32 320 32c-68.2 0-125 26.3-169.3 60.8L38.8 5.1zM223.1 149.5C248.6 126.2 282.7 112 320 112c79.5 0 144 64.5 144 144c0 24.9-6.3 48.3-17.4 68.7L408 294.5c5.2-11.8 8-24.8 8-38.5c0-53-43-96-96-96c-2.8 0-5.6 .1-8.4 .4c5.3 9.3 8.4 20.1 8.4 31.6c0 10.2-2.4 19.8-6.6 28.3l-90.3-70.8zm223.1 298L373 389.9c-16.4 6.5-34.3 10.1-53 10.1c-79.5 0-144-64.5-144-144c0-6.9 .5-13.6 1.4-20.2L83.1 161.5C60.3 191.2 44 220.8 34.5 243.7c-3.3 7.9-3.3 16.7 0 24.6c14.9 35.7 46.2 87.7 93 131.1C174.5 443.2 239.2 480 320 480c47.8 0 89.9-12.9 126.2-32.5z"/></svg></span></template>
        <template id=fa-copy><span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M502.6 70.63l-61.25-61.25C435.4 3.371 427.2 0 418.7 0H255.1c-35.35 0-64 28.66-64 64l.0195 256C192 355.4 220.7 384 256 384h192c35.2 0 64-28.8 64-64V93.25C512 84.77 508.6 76.63 502.6 70.63zM464 320c0 8.836-7.164 16-16 16H255.1c-8.838 0-16-7.164-16-16L239.1 64.13c0-8.836 7.164-16 16-16h128L384 96c0 17.67 14.33 32 32 32h47.1V320zM272 448c0 8.836-7.164 16-16 16H63.1c-8.838 0-16-7.164-16-16L47.98 192.1c0-8.836 7.164-16 16-16H160V128H63.99c-35.35 0-64 28.65-64 64l.0098 256C.002 483.3 28.66 512 64 512h192c35.2 0 64-28.8 64-64v-32h-47.1L272 448z"/></svg></span></template>
        <template id=fa-play><span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 384 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M73 39c-14.8-9.1-33.4-9.4-48.5-.9S0 62.6 0 80V432c0 17.4 9.4 33.4 24.5 41.9s33.7 8.1 48.5-.9L361 297c14.3-8.7 23-24.2 23-41s-8.7-32.2-23-41L73 39z"/></svg></span></template>
        <template id=fa-clock-rotate-left><span class=fa-svg><svg xmlns="http://www.w3.org/2000/svg" viewBox="0 0 512 512"><!--! Font Awesome Free 6.2.0 by @fontawesome - https://fontawesome.com License - https://fontawesome.com/license/free (Icons: CC BY 4.0, Fonts: SIL OFL 1.1, Code: MIT License) Copyright 2022 Fonticons, Inc. --><path d="M75 75L41 41C25.9 25.9 0 36.6 0 57.9V168c0 13.3 10.7 24 24 24H134.1c21.4 0 32.1-25.9 17-41l-30.8-30.8C155 85.5 203 64 256 64c106 0 192 86 192 192s-86 192-192 192c-40.8 0-78.6-12.7-109.7-34.4c-14.5-10.1-34.4-6.6-44.6 7.9s-6.6 34.4 7.9 44.6C151.2 495 201.7 512 256 512c141.4 0 256-114.6 256-256S397.4 0 256 0C185.3 0 121.3 28.7 75 75zm181 53c-13.3 0-24 10.7-24 24V256c0 6.4 2.5 12.5 7 17l72 72c9.4 9.4 24.6 9.4 33.9 0s9.4-24.6 0-33.9l-65-65V152c0-13.3-10.7-24-24-24z"/></svg></span></template>



        <script>
            window.playground_copyable = true;
        </script>


        <script src="elasticlunr-ef4e11c1.min.js"></script>
        <script src="mark-09e88c2c.min.js"></script>
        <script src="searcher-c2a407aa.js"></script>

        <script src="clipboard-1626706a.min.js"></script>
        <script src="highlight-abc7f01d.js"></script>
        <script src="book-a0b12cfe.js"></script>

        <!-- Custom JS scripts -->



    </div>
    </body>
</html>
